File: LAB6/LAB6_ejemplo.r
require(Matrix)
require(pracma)

#Funcion para generar el espacio nulo
EspacioNulo <- function (A) {
  m <- dim(A)[1]; n <- dim(A)[2]
  QR <- base::qr.default(A)
  r <- QR$rank
  if ((r < min(m, n)) || (m < n)) {
    R <- QR$qr[1:r, , drop = FALSE]
    P <- QR$pivot
    F <- R[, (r + 1):n, drop = FALSE]
    I <- base::diag(1, n - r)
    B <- -1.0 * base::backsolve(R, F, r)
    Y <- base::rbind(B, I)
    X <- Y[base::order(P), , drop = FALSE]
    return(X)
  }
  return(base::matrix(0, n, 1))
}




A <- matrix(c(1,2,1,  3,9,0,  1,4,-1,  4,9,3),3,4) 
A

#Encontrar el rango
RangoA <- rankMatrix(A)[1]
RangoA

#Encontrar la nulidad
nulidadA <- ncol(A)-RangoA
nulidadA

#Encontrar el espacio nulo
EspacioNuloA <- EspacioNulo(A)
EspacioNuloA

File: LAB6/LAB6_ejemplo.py
import numpy as np
import sympy as sp
from scipy.linalg import null_space


A = sp.Matrix([[1, 3, 1, 4],
               [2, 9, 4, 9],
               [1, 0, -1, 3]])

print(A)

# Encontrar el rango
RangoA = np.linalg.matrix_rank(np.array(A).astype(np.float64))
print("El rango de A es: "+str(RangoA))

# Encontrar la nulidad
num_rows, num_cols = np.array(A).astype(np.float64).shape
nulidadA = num_cols-RangoA
print("La nulidad de A es: " + str(nulidadA))

# Encontrar el espacio Nulo
EspacioNuloA = A.nullspace()
print("El espacio nulo de A es: \n "+str(EspacioNuloA))
